### PR TITLE
GCC-PHAT

### PR DESCRIPTION
**Pull request by luca-bondi**
GCC-PHAT added as feature option in `specgram`

The implementation is compatible with an arbitrary number of channels, and computes the GCC for each combination of input channels.

Needs to be tested with the TF version currently in use for the project, not sure what it is.

The provided `multich-sample.flac` sample contains a 4 channel recording with a linear microphone array of a car passing by. The GCC between channel 0 and channel 1 shows the delay changing over time:
gcc

### CODE REVIEW

File: ots_embedding/layers.py [hunk 1]
```diff
@@ -98,12 +99,53 @@ def call(x_input):
 
         if include_phase:
             y = Concatenate(axis=3)([y, Phase()(stft)])
+
+        if include_gcc:
+            # GCC-PHAT Cross-Correlation between channel pairs (http://www.xavieranguera.com/phdthesis/node92.html)
+            _, num_ch, _ = x_input.shape
```
Comment (magdalenafuentes) [inline ots_embedding/layers.py:105]: @beasteers double checking, `x_input.shape==(batch, num_ch, num_samples)`?

File: ots_embedding/layers.py [hunk 2]
```diff
@@ -98,12 +99,53 @@ def call(x_input):
...
+            if num_ch > 1:
+                # Need at least two channels for cross correlation to make sense
+                _, num_freq_coeff = y.shape[1:3]
```
Comment (magdalenafuentes) [inline ots_embedding/layers.py:108]: Sorry if it's a dumb question, but assuming `y.shape` is `(batch, time, freq)` why are you keeping `1:3` again?

File: ots_embedding/layers.py [hunk 3]
```diff
@@ -98,12 +99,53 @@ def call(x_input):
...
+                gcc_list = []
+                for ch0 in range(num_ch - 1):
+                    for ch1 in range(ch0 + 1, num_ch):
```
Comment (magdalenafuentes) [inline ots_embedding/layers.py:111]: About this, we were unsure with @beasteers if you can use `for loops` inside a layer?

**Review by magdalenafuentes — COMMENTED** [review]
@luca-bondi this is fantastic, thanks for putting it together so fast :)

I'm ccing @beasteers since she was thinking about this implementation as well and probably can help debug

File: ots_embedding/layers.py (hunk 2 above)
Comment (luca-bondi) [reply ots_embedding/layers.py:108]: good point, leftover from
`num_time_frames, num_freq_coeff = y.shape[1:3]`

I'll change to:
`num_freq_coeff = y_shape[2]`

**Comment by magdalenafuentes** [comment]
Another thing we discussed with @beasteers and it's worth giving a thought is the output dimension of `gcc`. Should it match the first dimension of the `stft` so it can be concatenated to it and fit to the `cnns` as a three channel tensor? That means `dim=257`, but @luca-bondi mentioned that it might be too noisy to keep that many coefficients for the `gcc`? Thoughts?

**Comment by luca-bondi** [comment]
> Another thing we discussed with @beasteers and it's worth giving a thought is the output dimension of `gcc`. Should it match the first dimension of the `stft` so it can be concatenated to it and fit to the `cnns` as a three channel tensor? That means `dim=257`, but @luca-bondi mentioned that it might be too noisy to keep that many coefficients for the `gcc`? Thoughts?

@magdalenafuentes I wouldn't worry too much about this point. As you see from the image above, the GCC component is basically low power noise when there is nothing interesting, and has high(yellow) or low(blue) values where there is some information. My intuition is this should play nicely with CNN training, as the areas of noise should be learned as not relevant by the network.

In the example above we are preserving 257 coefficients from the GCC to make the concatenation with the logspectrogram possible along the channel axis in

 https://github.com/magdalenafuentes/ots_embedding/blob/77e53f6b2c3f5a659bcdcab239e45f56270fd0f2/ots_embedding/layers.py#L140

 Despite the noise we can see the trace of the passage quite nicely!

**Comment by beasteers** [comment]
@luca-bondi this is looking great thanks!! 

I refactored your additions into their own GCC layer so that the model summary is more clear and it's a bit more reusable, but I verified that the output indeed looks like the plot you shared so yay!!

image

So I'd say this LGTM and we can go ahead and merge!

File: ots_embedding/layers.py (hunk 3 above)
Comment (beasteers) [reply ots_embedding/layers.py]: So it looks like this is fine with eager execution in tf 2 so I guess it's not an issue.

File: ots_embedding/layers.py (hunk 1 above)
Comment (beasteers) [reply ots_embedding/layers.py]: yep!

**Comment by magdalenafuentes** [comment]
Great, thanks both! I’m merging this then

**Pull request merged**